docs/RFCS/20200810_leaseholder_cache_invalidation.md

```diff
@@ -0,0 +1,464 @@
+  the space complexity from O(#ranges) to O(#active ranges).
+
+- Due to the eventual consistency of gossip, there's an potentially
+  long amount of time between the moment a node terminates ([step 5 of
+  drain](#draining)) and the moment other nodes learn of the drain
+  record. During that time, requests can be mis-routed.
+
+  This last cons can be alleviated by extending step 5 of the draining
+  process by a "sleep period" to let gossip propagate, calibrated to
+  the standard cross-cluster gossip delivery latency (for reference,
+  that's around 10-20s for up to 100 geo-distributed nodes in normal
+  operation).
+
+### Multi-tenant variant
+
+- The KV servers receive the "drain records" over gossip from other KV servers.
+- The KV servers *forward* the drain records to SQL tenants,
+  where the range caches are stored.
+
+## 3. Gossiping lease acquisitions
+
+Outline:
+
+- each node acquiring a lease (either because it was given to it
+  during another's node drain, or it is voted to be when a node goes
+  AWOL) gossips this fact in a "lease record", keyed by the range ID
+  with the lease as payload.
+- upon receiving "lease records", other nodes update their cache
+  with the gossiped lease, if it's newer than what they know.
+
+This approach was prototyped here: https://github.com/cockroachdb/cockroach/pull/52572
+
+**Pros:**
+
+- Addresses both [scenarios A and B](#scenarios).
+- By using gossip, the data will eventually reach every other
+  node even those that were temporarily unavailable.
+  (i.e. eliminates one cons of solution 1 above)
+- By informing the other nodes of the new leaseholder,
+  this solution accelerates/removes the need for the
+  lease discovery mechanism.
+  (i.e. eliminates the other cons of solution 1 above)
+
+**Cons:**
+
+- There is one "lease record" in gossip per range.
+  This violates the [design principles](#general-cockroachdb-design-principles).
+
+  We can partially alleviate this cost by populating the lease expiry
+  time as the value of [gossip TTL](#gossip). This reduces
+  the space complexity from O(#ranges) to O(#active ranges).
+
+- Like above, there's a delay between the moment a node terminates
+  ([step 5 of drain](#draining)) and the moment other nodes learn of
+  the new lease records. During that time, requests can be mis-routed.
+
+  Like above, this last cons can be alleviated by extending step 5 of
+  the draining process by a "sleep period".
+
+### Multi-tenant variant
+
+- The KV servers receive the "lease records" over gossip from other KV servers.
+- The KV servers *forward* the lease records to SQL tenants,
+  where the range caches are stored.
+
+## 4. Liveness-based cache invalidation with poison
+
+Outline:
+
+- [liveness updates](#liveness-updates) are gossiped as usual.
+- upon receiving a liveness update that indicates that another node is
+  not live, each node updates its range cache as follows:
+
+  - invalidate all leases pointing to the now-unavailable node, ie.
```

Comment:
You don't necessarily even need to update the cache in order to take advantage of a liveness update; since we generally have full leases in the cache, a DistSender knows the epoch of each lease. If it's lower than the liveness info, then it knows that that lease is not valid (that another reason why the cache maintains full lease info and also why we update the cache even on successful RPCs when the client had stale info).

So I don't think you need code to explicitly deal with any "poisoning"; I think between the node info telling you if the cache is stale, and the liveness record expiration timestamp and also the connection health info exposed by the RPC context, we have enough to determine both if we trust a lease and if a node currently looks to be up or down (and thus should be de-prioritized for discovery).

later update: I see now option 5. Let's get rid of 4 then; it's dominated isn't it?